gst/cgo_exports.go

```diff
@@ -261,15 +261,16 @@ func goLogFunction(
 	if f != nil {
 		var obj *glib.Object
 		if object != nil {
-			obj = glib.TransferNone(unsafe.Pointer(object))
+			obj = glib.TransferNone(unsafe.Pointer(object)).Ref()
```

Comment:
I see from your commit messages, that this fixes a crash, but I think this way you are leaking memory. Every time an object logs a message you ref it without unreffing it.

Reply:
That was my original assumption as well, but let me open this up a bit, maybe there's a bigger issue somewhere underneath glib internals to look into?

I run an app that uses webrtcsrc/webrtcsink for video streaming related things and I noticed that as soon as I enable custom logging at all, negotiating the pipeline usually stops working at all.

```
(serviced:151270): GStreamer-WebRTC-CRITICAL **: 02:14:05.522: gst_webrtc_ice_set_is_controller: assertion 'GST_IS_WEBRTC_ICE (ice)' failed
(serviced:151270): GLib-GObject-CRITICAL **: 02:14:05.522: g_object_get: assertion 'G_IS_OBJECT (object)' failed
(serviced:151270): GLib-GObject-CRITICAL **: 02:14:05.522: g_object_get: assertion 'G_IS_OBJECT (object)' failed
(serviced:151270): GLib-GObject-CRITICAL **: 02:14:05.523: g_object_get: assertion 'G_IS_OBJECT (object)' failed
(serviced:151270): GLib-GObject-CRITICAL **: 02:14:05.523: g_object_get: assertion 'G_IS_OBJECT (object)' failed
(serviced:151270): GStreamer-WebRTC-CRITICAL **: 02:14:05.523: gst_webrtc_ice_set_local_credentials: assertion 'GST_IS_WEBRTC_ICE (ice)' failed
(serviced:151270): GStreamer-WebRTC-CRITICAL **: 02:14:05.523: gst_webrtc_ice_gather_candidates: assertion 'GST_IS_WEBRTC_ICE (ice)' failed
```

I wasn't using the glib.Object passed to logging callback at all, but alright, I added something along of
```
if object != nil {
  object.Ref()
}

```

Into my callback and it fixed the issue. Some of my debug levels use debug/trace levels, so there can be some traffic, so I kept monitoring the memory utilization for few hours and didn't seem to be a problem.

After trying to add `Unref()` to the end of the logging callback function though, I started to get crashes again.

Reply:
> so I kept monitoring the memory utilization for few hours and didn't seem to be a problem.

I think this depends on your pipeline. If you create your pipeline once and let it run, then this should be fine (you create gigantic refcounts but no memory leaks). If you instead have a dynamic pipeline then removing old elements is "impossible", because you don't know how many logs the elements put out.

I'm not so sure about this so I asked the gstreamer devs for help.

Reply:
I got a response from @sdroege:

> you must not ref_sink() whatever object is passed into the logging function, or you'll take ownership of random objects that happen to be logged before someone else takes ownership of them
> 
> you must not change the object's state at all
> also, objects passed into the logging function might be not fully constructed yet (logging before g_object_new() has returned), or they might have a refcount of 0 (logging in finalize), or they might be partially destructed (dispose)
> that's why the rust bindings don't provide a proper full object there https://gstreamer.pages.freedesktop.org/gstreamer-rs/stable/latest/docs/gstreamer/log/fn.add_log_function.html . but just something that can be printed, and otherwise gives access to a raw pointer if you know what you're doing
> 
>  generally, for vfuncs/signal handlers/callbacks you want to make sure to not ref_sink() object references that are passed as transfer-none


So the reason for your errors @aet is the `TransferNone`, not the missing `Ref`. Since the current implementation is "broken", I would be also fine with breaking its signature.

Reply:
Ah this makes perfect sense now and definitely better than my naive first fix attempt. I added some fixes that are following what `gstreamer-rs` does with some naming conventions, and checking for `category == nil` like the rust version does. I suppose it was related to some special case of gstreamer builds having no debugging enabled.